plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java

```diff
@@ -123,30 +144,144 @@ public DeltaLakeMergeSink(
+        }
+
+        int positionCount = inputPage.getPositionCount();
+        if (positionCount <= 0) {
+            throw new IllegalArgumentException("positionCount should be > 0, but is " + positionCount);
+        }
+        Block operationBlock = inputPage.getBlock(inputChannelCount - 2);
+        int[] deletePositions = new int[positionCount];
+        int[] insertPositions = new int[positionCount];
+        int[] updateInsertPositions = new int[positionCount];
+        int[] updateDeletePositions = new int[positionCount];
+        int deletePositionCount = 0;
+        int insertPositionCount = 0;
+        int updateInsertPositionCount = 0;
+        int updateDeletePositionCount = 0;
+
+        for (int position = 0; position < positionCount; position++) {
+            int operation = toIntExact(TINYINT.getLong(operationBlock, position));
+            switch (operation) {
+                case DELETE_OPERATION_NUMBER:
+                    deletePositions[deletePositionCount] = position;
+                    deletePositionCount++;
+                    break;
+                case INSERT_OPERATION_NUMBER:
+                    insertPositions[insertPositionCount] = position;
+                    insertPositionCount++;
+                    break;
+                case UPDATE_INSERT_OPERATION_NUMBER:
+                    updateInsertPositions[updateInsertPositionCount] = position;
+                    updateInsertPositionCount++;
+                    break;
+                case UPDATE_DELETE_OPERATION_NUMBER:
+                    updateDeletePositions[updateDeletePositionCount] = position;
+                    updateDeletePositionCount++;
+                    break;
+                default:
+                    throw new IllegalArgumentException("Invalid merge operation: " + operation);
             }
-        });
+        }
+        Optional<Page> deletePage = Optional.empty();
+        if (deletePositionCount > 0) {
+            int[] columns = new int[dataColumnCount + 1];
+            for (int i = 0; i < dataColumnCount; i++) {
+                columns[i] = i;
+            }
+            columns[dataColumnCount] = dataColumnCount + 1; // row ID channel
+            deletePage = Optional.of(inputPage
+                    .getColumns(columns)
+                    .getPositions(deletePositions, 0, deletePositionCount));
+        }
+
+        Optional<Page> insertPage = Optional.empty();
+        if (insertPositionCount > 0) {
+            insertPage = Optional.of(inputPage
+                    .getColumns(IntStream.range(0, dataColumnCount).toArray())
+                    .getPositions(insertPositions, 0, insertPositionCount));
+        }
+
+        Optional<Page> updateInsertPage = Optional.empty();
+        if (updateInsertPositionCount > 0) {
+            updateInsertPage = Optional.of(inputPage
+                    .getColumns(IntStream.range(0, dataColumnCount).toArray())
```

Comment:
Above you do this with ordinary for loop

```
int[] columns = new int[dataColumnCount + 1];
            for (int i = 0; i < dataColumnCount; i++) {
                columns[i] = i;
            }
```

we generally avoid streams in "execution" because they have some small overhead
anyway, `dataColumnCount` is constant, so the column indices arrays could be pre-computed in the ctor (and then it they can be stream-based, if that is more readable)